psydac/linalg/block.py

```diff
@@ -533,7 +534,7 @@ def conjugate(self, out=None):
         else:
             out = BlockLinearOperator(self.domain, self.codomain)
 
-        for (i, j), Lij in self._blocks_as_args.items():
+        for (i, j), Lij in self._blocks.items():
```

Comment:
What is the difference between `_blocks` and `_blocks_as_args`?

Reply:
For some reason, the `BlockLinearOperator._dot` method takes `self._blocks_as_args` as argument and is able to use `.items()` on it. See devel branch, block.py lines 610, 617 and and 621.

Reply:
Exactly, in the `__init__` of `BlockLinearOperator` line `496` it says `self._blocks_as_args = self._blocks`

Reply:
In fact, `self._blocks_as_args` is define as `self_blocks` in the initialization of a `BlockLinearOperator`. But when you change the `backend` of this object, `self._blocks_as_args` is then define as a list of element of the elements of `self.blocks`. So, if you change the backend, you change the `self.func` and thus you don't use `self._dot` anymore. 

But, the method `conjugate` didn't change with the backend, thus we should have in this case `self._blocks` and not `self._blocks_as_args`

Reply:
Thanks for the input!

From what I understand `self._blocks_as_args` is used when code is generated for a `BlockLinearOperator` containing only objects of type `StencilMatrix` and `StencilInterfaceMatrix`. I suspect that this kind of code generation is not fully necessary given the recent changes removing temporaries from`BlockLinearOperator.dot`.

Having an accelerated `idot` function would make `BlockLinearOperator.dot` even more efficient as we would minimize the amount read/write memory operations, but I think that the performance of the new solution should not be too far behind.

@saidctb Does that sound correct to you? Could you please compare the accelerated `BlockLinearOperator.dot` (i.e. the old solution) with the one where the dot products of each block are accelerated (new solution)?

Reply:
Yes, this sound correct.
Testing the `BlockLinearOperator.dot` with the example in `test_api_2d_system.py` with `ncells=[2**3, 2**3]` and `p=3` I got.
```
dot product timing with each block accelerated separately we get: 8.680629730224609e-05 s
dot product timing with the whole block accelerated we get: 7.189559936523438e-05 s
```

Reply:
The solution with each block accelerated separately is quite good already! The domain is very small though, could you try something bigger?